### PR TITLE
Prevent payload config file overwrite

### PR DESCRIPTION
**Pull request by uruwhy**
## Description
Previously, when loading payload config files from `plugins/[plugin_name]/data/payloads/*.yml`, the data service would overwrite the existing payload config each time it loads a new config yaml, so that only the last seen yaml gets saved. This PR addresses this bug by making sure all loaded yamls get combined into one since config yaml.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
Wrote unit test to address the changes. Also tested with actual config yamls in two different plugins and confirmed that both configurations were combined and that the second one didn't overwrite the first. The test will also override the payload config from the first plugin, and I confirmed that the override occurred and that the logger produces the expected warning output (checked via `python3 -m pytest tests/services --log-cli-level=WARN`):
```
------------------------------------------------------------- live log call -------------------------------------------------------------
WARNING  data_svc:data_svc.py:359 Config for file.cpp already exists in the special_payloads section of the payloads config and will be overridden by plugin virtual
```

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tests/services/test_data_svc.py [hunk 1]
```diff
@@ -92,3 +167,72 @@ def test_no_autogen_cleanup_cmds(self, event_loop, data_svc):
...
+                'file.go': PAYLOAD_CONFIG_YAMLS['path1'][0]['special_payloads']['file.go'],
+                'file.cpp': PAYLOAD_CONFIG_YAMLS['path2'][0]['special_payloads']['file.cpp'],
+                'special.py': PAYLOAD_CONFIG_YAMLS['path3'][0]['special_payloads']['special.py'],
+            },
+            'extensions': {
+                '.donut': PAYLOAD_CONFIG_YAMLS['path1'][0]['extensions']['.donut'],
+                '.testext': PAYLOAD_CONFIG_YAMLS['path3'][0]['extensions']['.testext'],
+            }
+        }
+        mock_apply_config2.assert_called_once_with(name='payloads', config=expected_config_part2)
```
Comment (clenk) [inline tests/services/test_data_svc.py:238]: Do we need to test what happens if two plugins have the same payload but maybe a different id?

File: app/service/data_svc.py [hunk 2]
```diff
@@ -310,15 +314,31 @@ async def _load_objectives(self, plugin):
...
         for filename in glob.iglob('%s/payloads/*.yml' % plugin.data_dir, recursive=False):
             data = self.strip_yml(filename)
-            payload_config = self.get_config(name='payloads')
-            payload_config['standard_payloads'] = data[0]['standard_payloads']
-            payload_config['special_payloads'] = data[0]['special_payloads']
-            payload_config['extensions'] = data[0]['extensions']
-            await self._apply_special_payload_hooks(payload_config['special_payloads'])
-            await self._apply_special_extension_hooks(payload_config['extensions'])
-            self.apply_config(name='payloads', config=payload_config)
+            special_payloads = data[0].get(PAYLOADS_CONFIG_SPECIAL_KEY, dict())
```
[the remark this answers is not in the document]
Comment (uruwhy) [reply app/service/data_svc.py:324]: `special_payloads` and `extensions` get passed as arguments in multiple functions, so that's why I pulled them out into variables. `standard_payloads1 only gets used once in the `update` call, so I didn't bother pulling it out there.

File: app/service/data_svc.py [hunk 3]
```diff
@@ -310,15 +314,31 @@ async def _load_objectives(self, plugin):
...
+        self._update_payload_config(payload_config)
+
+    def _update_payload_config(self, updates):
+        payload_config = self.get_config(name='payloads')
+        payload_config[PAYLOADS_CONFIG_STANDARD_KEY] = {**payload_config.get(PAYLOADS_CONFIG_STANDARD_KEY, dict()),
+                                                        **updates.get(PAYLOADS_CONFIG_STANDARD_KEY, dict())}
+        payload_config[PAYLOADS_CONFIG_SPECIAL_KEY] = {**payload_config.get(PAYLOADS_CONFIG_SPECIAL_KEY, dict()),
+                                                       **updates.get(PAYLOADS_CONFIG_SPECIAL_KEY, dict())}
+        payload_config[PAYLOADS_CONFIG_EXTENSIONS_KEY] = {**payload_config.get(PAYLOADS_CONFIG_EXTENSIONS_KEY, dict()),
+                                                          **updates.get(PAYLOADS_CONFIG_EXTENSIONS_KEY, dict())}
```
[the remark this answers is not in the document]
Comment (uruwhy) [reply app/service/data_svc.py:340]: Yeah, I'd like to default to empty dicts where possible just in case

File: tests/services/test_data_svc.py (hunk 1 above)
Comment (uruwhy) [reply tests/services/test_data_svc.py:238]: I haven't considered that edge case - what do we want as the outcome? The most recently processed plugin overrides previous entries?

File: app/service/data_svc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clenk) [reply app/service/data_svc.py:324]: Not a big deal but could improve readability a tiny bit.

File: tests/services/test_data_svc.py (hunk 1 above)
Comment (clenk) [reply tests/services/test_data_svc.py:238]: Will "most recently processed" be consistent? Are they in alphabetical order or random?

Probably also want to issue an INFO or DEBUG message about it, to help point users to what happened if they expected one thing but got another.

File: app/service/data_svc.py [hunk 4]
```diff
@@ -310,15 +314,31 @@ async def _load_objectives(self, plugin):
...
-            payload_config['standard_payloads'] = data[0]['standard_payloads']
-            payload_config['special_payloads'] = data[0]['special_payloads']
-            payload_config['extensions'] = data[0]['extensions']
-            await self._apply_special_payload_hooks(payload_config['special_payloads'])
-            await self._apply_special_extension_hooks(payload_config['extensions'])
-            self.apply_config(name='payloads', config=payload_config)
+            special_payloads = data[0].get(PAYLOADS_CONFIG_SPECIAL_KEY, dict())
+            extensions = data[0].get(PAYLOADS_CONFIG_EXTENSIONS_KEY, dict())
+            await self._apply_special_payload_hooks(special_payloads)
+            await self._apply_special_extension_hooks(extensions)
```
Comment (christophert) [inline app/service/data_svc.py:327]: IMO, We should only call these processing functions and `payload_config` updates on the respective data if the respective `dicts` are not empty

**Review by christophert — CHANGES_REQUESTED** [review]

File: app/service/data_svc.py (hunk 4 above)
Comment (uruwhy) [reply app/service/data_svc.py:327]: Those functions don't do anything if the dict is empty, but I can add the checks in here anyway

File: app/service/data_svc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (uruwhy) [reply app/service/data_svc.py:324]: done

File: tests/services/test_data_svc.py (hunk 1 above)
Comment (uruwhy) [reply tests/services/test_data_svc.py:243]: Added some checks and warns before override. Adjusted test to include an override example

File: app/service/data_svc.py [hunk 5]
```diff
@@ -323,23 +323,42 @@ async def _load_payloads(self, plugin):
...
+
+        self._check_payload_overlaps(curr_standard_payloads, new_standard_payloads, PAYLOADS_CONFIG_STANDARD_KEY,
+                                     curr_plugin_name)
+        self._check_payload_overlaps(curr_special_payloads, new_special_payloads, PAYLOADS_CONFIG_SPECIAL_KEY,
+                                     curr_plugin_name)
+        self._check_payload_overlaps(curr_extensions, new_extensions, PAYLOADS_CONFIG_EXTENSIONS_KEY, curr_plugin_name)
+
+        payload_config[PAYLOADS_CONFIG_STANDARD_KEY] = {**curr_standard_payloads, **new_standard_payloads}
+        payload_config[PAYLOADS_CONFIG_SPECIAL_KEY] = {**curr_special_payloads, **new_special_payloads}
+        payload_config[PAYLOADS_CONFIG_EXTENSIONS_KEY] = {**curr_extensions, **new_extensions}
```
Comment (clenk) [inline app/service/data_svc.py:353]: Could use `.update()` instead, because of how the code is structured now.

**Review by clenk — APPROVED** [review]
Looks good to me - not sure why Sonarcloud failed or how to rerun it.

File: app/service/data_svc.py (hunk 5 above)
Comment (uruwhy) [reply app/service/data_svc.py:353]: `.update()` modifies in place, whereas `{**,**}` will give us a new dict, which allows me to handle cases where maybe `payload_config` doesn't yet have a particular section. Like if `payload_config[PAYLOADS_CONFIG_EXTENSIONS_KEY]` returns `None`, I don't want to call `.update()` on that, and it'll get too verbose to add checks there.

**Review by clenk — APPROVED** [review]
LGTM!
